### ISSUE TITLE
Inacurrate behavior in case of a topic or broker misconfiguration

### ISSUE THREAD

**Issue by teivah** [beatlabs/patron#443]
## Is your feature request related to a problem? Please describe

We noticed an inaccurate behavior when creating a Kafka consumer in the following cases:
* Topic name is misconfigured and pointing to an unexisting topic (and auto-create is disabled)
* The broker addresses are misconfigured

In both cases, Patron is going to set both the liveness to true but handle these errors as retries. 

The problem we faced in our deployment went ok, and the service started. Yet, after x minutes (x = period * nb of retries), Patron exhausted the number of retries and stopped the service.

It led to an incident because we had three replicas that started more or less at the same time and which all stopped also at the same time. Hence, at some point, we went down to zero replicas.

## Is this a bug? Please provide steps to reproduce, a failing test, etc.

Inexisting topic or wrong broker configuration

## Describe the solution

In my opinion, I would prefer Patron not to set the service as live & ready in the two cases described. It's probably okay to go for retries (at least in the case of the broker being unreachable because we don't know for sure if it's a misconfiguration or a transient error). However, we shouldn't set the service as ready and live as long as we have been able to create all the Kafka consumers.

**Comment by rbroggi** [comment]
I'm taking this ticket. Before proceeding I'm gathering all the kafka errors that I think could be classified as 'unrecoverable' -> topic not existing and broker not existing (e.g.: `ConfigurationError` is for sure something that is not recoverable, therefore I would also include this and the other list of errors I'm collecting in the scope of this change). 

On top of the `ConfigurationError` I'm assessing the following list to determine what should be considered 'unrecoverable' -> therefore something that should not be retried from what should be considered 'recoverable'-> therefore should be retried:

```go
const (
	ErrNoError                            KError = 0
	ErrUnknown                            KError = -1
	ErrOffsetOutOfRange                   KError = 1
	ErrInvalidMessage                     KError = 2
	ErrUnknownTopicOrPartition            KError = 3
	ErrInvalidMessageSize                 KError = 4
	ErrLeaderNotAvailable                 KError = 5
	ErrNotLeaderForPartition              KError = 6
	ErrRequestTimedOut                    KError = 7
	ErrBrokerNotAvailable                 KError = 8
	ErrReplicaNotAvailable                KError = 9
	ErrMessageSizeTooLarge                KError = 10
	ErrStaleControllerEpochCode           KError = 11
	ErrOffsetMetadataTooLarge             KError = 12
	ErrNetworkException                   KError = 13
	ErrOffsetsLoadInProgress              KError = 14
	ErrConsumerCoordinatorNotAvailable    KError = 15
	ErrNotCoordinatorForConsumer          KError = 16
	ErrInvalidTopic                       KError = 17
	ErrMessageSetSizeTooLarge             KError = 18
	ErrNotEnoughReplicas                  KError = 19
	ErrNotEnoughReplicasAfterAppend       KError = 20
	ErrInvalidRequiredAcks                KError = 21
	ErrIllegalGeneration                  KError = 22
	ErrInconsistentGroupProtocol          KError = 23
	ErrInvalidGroupId                     KError = 24
	ErrUnknownMemberId                    KError = 25
	ErrInvalidSessionTimeout              KError = 26
	ErrRebalanceInProgress                KError = 27
	ErrInvalidCommitOffsetSize            KError = 28
	ErrTopicAuthorizationFailed           KError = 29
	ErrGroupAuthorizationFailed           KError = 30
	ErrClusterAuthorizationFailed         KError = 31
	ErrInvalidTimestamp                   KError = 32
	ErrUnsupportedSASLMechanism           KError = 33
	ErrIllegalSASLState                   KError = 34
	ErrUnsupportedVersion                 KError = 35
	ErrTopicAlreadyExists                 KError = 36
	ErrInvalidPartitions                  KError = 37
	ErrInvalidReplicationFactor           KError = 38
	ErrInvalidReplicaAssignment           KError = 39
	ErrInvalidConfig                      KError = 40
	ErrNotController                      KError = 41
	ErrInvalidRequest                     KError = 42
	ErrUnsupportedForMessageFormat        KError = 43
	ErrPolicyViolation                    KError = 44
	ErrOutOfOrderSequenceNumber           KError = 45
	ErrDuplicateSequenceNumber            KError = 46
	ErrInvalidProducerEpoch               KError = 47
	ErrInvalidTxnState                    KError = 48
	ErrInvalidProducerIDMapping           KError = 49
	ErrInvalidTransactionTimeout          KError = 50
	ErrConcurrentTransactions             KError = 51
	ErrTransactionCoordinatorFenced       KError = 52
	ErrTransactionalIDAuthorizationFailed KError = 53
	ErrSecurityDisabled                   KError = 54
	ErrOperationNotAttempted              KError = 55
	ErrKafkaStorageError                  KError = 56
	ErrLogDirNotFound                     KError = 57
	ErrSASLAuthenticationFailed           KError = 58
	ErrUnknownProducerID                  KError = 59
	ErrReassignmentInProgress             KError = 60
	ErrDelegationTokenAuthDisabled        KError = 61
	ErrDelegationTokenNotFound            KError = 62
	ErrDelegationTokenOwnerMismatch       KError = 63
	ErrDelegationTokenRequestNotAllowed   KError = 64
	ErrDelegationTokenAuthorizationFailed KError = 65
	ErrDelegationTokenExpired             KError = 66
	ErrInvalidPrincipalType               KError = 67
	ErrNonEmptyGroup                      KError = 68
	ErrGroupIDNotFound                    KError = 69
	ErrFetchSessionIDNotFound             KError = 70
	ErrInvalidFetchSessionEpoch           KError = 71
	ErrListenerNotFound                   KError = 72
	ErrTopicDeletionDisabled              KError = 73
	ErrFencedLeaderEpoch                  KError = 74
	ErrUnknownLeaderEpoch                 KError = 75
	ErrUnsupportedCompressionType         KError = 76
	ErrStaleBrokerEpoch                   KError = 77
	ErrOffsetNotAvailable                 KError = 78
	ErrMemberIdRequired                   KError = 79
	ErrPreferredLeaderNotAvailable        KError = 80
	ErrGroupMaxSizeReached                KError = 81
	ErrFencedInstancedId                  KError = 82
	ErrEligibleLeadersNotAvailable        KError = 83
	ErrElectionNotNeeded                  KError = 84
	ErrNoReassignmentInProgress           KError = 85
	ErrGroupSubscribedToTopic             KError = 86
	ErrInvalidRecord                      KError = 87
	ErrUnstableOffsetCommit               KError = 88
)
```


Here is a first iteration:

```
ErrNoError
ErrUnknown
ErrOffsetOutOfRange
ErrInvalidMessage
ErrUnknownTopicOrPartition -> (Unrecoverable)
ErrInvalidMessageSize
ErrLeaderNotAvailable
ErrNotLeaderForPartition
ErrRequestTimedOut
ErrBrokerNotAvailable -> (Unrecoverable)
ErrReplicaNotAvailable
ErrMessageSizeTooLarge
ErrStaleControllerEpochCode
ErrOffsetMetadataTooLarge
ErrNetworkException
ErrOffsetsLoadInProgress
ErrConsumerCoordinatorNotAvailable
ErrNotCoordinatorForConsumer
ErrInvalidTopic -> (Unrecoverable)
ErrMessageSetSizeTooLarge
ErrNotEnoughReplicas
ErrNotEnoughReplicasAfterAppend
ErrInvalidRequiredAcks
ErrIllegalGeneration
ErrInconsistentGroupProtocol
ErrInvalidGroupId -> (WDYT???)
ErrUnknownMemberId
ErrInvalidSessionTimeout
ErrRebalanceInProgress
ErrInvalidCommitOffsetSize
ErrTopicAuthorizationFailed -> (Unrecoverable)
ErrGroupAuthorizationFailed -> (Unrecoverable)
ErrClusterAuthorizationFailed -> (Unrecoverable)
ErrInvalidTimestamp
ErrUnsupportedSASLMechanism
ErrIllegalSASLState
ErrUnsupportedVersion -> (Unrecoverable)
ErrTopicAlreadyExists
ErrInvalidPartitions
ErrInvalidReplicationFactor
ErrInvalidReplicaAssignment
ErrInvalidConfig -> (Unrecoverable)
ErrNotController
ErrInvalidRequest
ErrUnsupportedForMessageFormat
ErrPolicyViolation -> (WDYT???)
ErrOutOfOrderSequenceNumber
ErrDuplicateSequenceNumber
ErrInvalidProducerEpoch
ErrInvalidTxnState
ErrInvalidProducerIDMapping
ErrInvalidTransactionTimeout
ErrConcurrentTransactions
ErrTransactionCoordinatorFenced
ErrTransactionalIDAuthorizationFailed
ErrSecurityDisabled
ErrOperationNotAttempted
ErrKafkaStorageError
ErrLogDirNotFound
ErrSASLAuthenticationFailed -> (Unrecoverable)
ErrUnknownProducerID
ErrReassignmentInProgress
ErrDelegationTokenAuthDisabled
ErrDelegationTokenNotFound
ErrDelegationTokenOwnerMismatch
ErrDelegationTokenRequestNotAllowed
ErrDelegationTokenAuthorizationFailed
ErrDelegationTokenExpired
ErrInvalidPrincipalType
ErrNonEmptyGroup
ErrGroupIDNotFound
ErrFetchSessionIDNotFound
ErrInvalidFetchSessionEpoch
ErrListenerNotFound
ErrTopicDeletionDisabled
ErrFencedLeaderEpoch
ErrUnknownLeaderEpoch
ErrUnsupportedCompressionType
ErrStaleBrokerEpoch
ErrOffsetNotAvailable
ErrMemberIdRequired
ErrPreferredLeaderNotAvailable
ErrGroupMaxSizeReached
ErrFencedInstancedId
ErrEligibleLeadersNotAvailable
ErrElectionNotNeeded
ErrNoReassignmentInProgress
ErrGroupSubscribedToTopic
ErrInvalidRecord
ErrUnstableOffsetCommit

```

**Comment by mantzas** [comment]
@rbroggi @teivah We discussed a different approach though, which was on the creation of the kafka component and the check if the topic exists or not. Did the decision change?

**Comment by teivah** [comment]
> @rbroggi @teivah We discussed a different approach though, which was on the creation of the kafka component and the check if the topic exists or not. Did the decision change?

@mantzas I don't think so. It's still about doing a check [here](https://github.com/teivah/patron/blob/ce98403c9d5b8dacde634f165d7ae928867813c9/component/kafka/group/component.go#L154) to see if the broker and the topic are available.

**Comment by mantzas** [comment]
@teivah then we are aligned. I though we would try based on the errors to identify a failing components. LEt's do it then @rbroggi  thanks

**Comment by rbroggi** [comment]
coll, that was exactly the spot I was investigating. Thx guys!

I will do it in the afternoon

**Comment by rbroggi** [comment]
After profitable discussion with @teivah and @mantzas we decided to avoid enriching directly the construction because:

1. In the constructor we have no 'context' and making a network call that is not cancellable is not ideal.
2. There is a risk we break client-library code by placing a network call in the constructor - it was not there before and library clients might be leveraging this on unit tests.

We have decided to use an `optionfunc` for this behavior to have 100% backward compliance while empowering interesting users to have the feature. 

We have also decided to reduce scope only to the `component/kafka` and disregard the legacy `component/async/kafka`.
